File: analisador.cpp
#include <iostream>
#include <string>
#include <stdio.h>
#include <ctype.h>

using namespace std;
string tape;
int index = 0;
string tempS;
string tabVar[500];
string output = "";
int tabVarLength = 0;
int tempN = 0;

void e0();
void e1();
void e2();

bool isBlank(){
    if(tape[index] == ' '){
        return true;
    }
    return false;
}

int contain(){
    int existe = -1;
    for (int i = 0; i < tabVarLength; i++) {
        if (tempS ==  tabVar[i]){
            existe = i;
            break;
        }
    }
    return existe;
}

void tabela(){
    for (int i = 0; i < tabVarLength; i++) {
        cout << i << " ... " << tabVar[i] << endl;
    }  
}

void sigma1(){
    tempS = tape[index];
}

void sigma2(){
    tempS = tempS + tape[index];   
}

void sigma3(){
    if (tabVarLength > 0){
        int aux = contain();
        if(aux >= 0){
            cout << "V(" << aux << ")";
        } else {
            tabVar[tabVarLength] = tempS;
            cout << "V(" << tabVarLength << ")";
            tabVarLength++;
        }
    } else {
        tabVar[tabVarLength] = tempS;
        cout << "V(" << tabVarLength << ")";
        tabVarLength++;
    }
}

void sigma4(){
    tempN = tape[index] - '0';
}

void sigma5(){
    tempN = (tape[index] - '0') + (tempN * 10);
}

void sigma6(){
    cout << "N(" << tempN << ")";
}

void e0(){
    tempS = "";
    tempN = 0;
    if(isBlank()){
        index++;
        e0();
    } else if(isdigit(tape[index])){
        sigma4();
        index++;
        e2();
    } else if(isalpha(tape[index])){
        sigma1();
        index++;
        e1();
    } else if (tape[index] == 0){
        cout << output << endl;
        tabela();
        cout << endl;
    } else {
        cout << "REJEITADO" << endl;
    }
}

void e1(){
    if(isdigit(tape[index]) || isalpha(tape[index])){
        sigma2();
        index++;
        e1();
    } else {
        sigma3();
        e0();
    }
}

void e2(){
    if(isdigit(tape[index])){
        sigma5();
        index++;
        e2();
    } else{
        sigma6();
        e0();
    }
}

int main() {
    cout << "Input the word: ";
    getline(cin, tape);
    e0();
    cin.get();
    return 0;
}

File: analisador.py
class Analisador(object):
    """Analisador lexico."""

    def __init__(self):
        self.tape = ''
        self.index = 0
        self.tempS = ''
        self.tabVar = []
        self.output = ''

    def isBlank(self):
        """."""
        return self.tape[self.index] == ' '

    def contain(self):
        """."""
        for index, value in enumerate(self.tabVar):
            if value == self.tempS:
                return index
        return -1

    def up_index(self):
        """."""
        self.index += 1

    def tabela(self):
        """."""
        for index, value in enumerate(self.tabVar):
            print('%s ... %s' % (index, value))

    def sigma1(self):
        """."""
        self.tempS = self.tape[self.index]

    def sigma2(self):
        """."""
        self.tempS = self.tempS + self.tape[self.index]

    def sigma3(self):
        """."""
        if len(self.tabVar) == 0:
            self.tabVar.append(self.tempS)
            self.output += 'V(%s)' % (len(self.tabVar) - 1)
        else:
            tmp = self.contain()
            if tmp >= 0:
                self.output += 'V(%s)' % (tmp)
            else:
                self.tabVar.append(self.tempS)
                self.output += 'V(%s)' % (len(self.tabVar) - 1)

    def sigma4(self):
        """."""
        self.tempN = int(self.tape[self.index])

    def sigma5(self):
        """."""
        self.tempN = (self.tempN * 10) + int(self.tape[self.index])

    def sigma6(self):
        """."""
        self.output += 'N(%s)' % (self.tempN)

    def e0(self):
        """."""
        if self.index == len(self.tape):
            print(self.output)
            self.tabela()
        elif(self.isBlank()):
            self.up_index()
            self.e0()
        elif self.tape[self.index].isdigit():
            self.sigma4()
            self.up_index()
            self.e2()
        elif self.tape[self.index].isalpha():
            self.sigma1()
            self.up_index()
            self.e1()
        else:
            raise Exception("REJEITADO")

    def e1(self):
        """."""
        if self.tape[self.index].isdigit() or self.tape[self.index].isalpha():
            self.sigma2()
            self.up_index()
            self.e1()
        else:
            self.sigma3()
            self.e0()

    def e2(self):
        """."""
        if self.tape[self.index].isdigit():
            self.sigma5()
            self.up_index()
            self.e2()
        else:
            self.sigma6()
            self.e0()

    def main(self):
        """."""
        self.tape = input("Digite a palavra: ")
        self.tape += ' '
        self.e0()
        input('')
        return

analisador = Analisador()
analisador.main()
